scio-core/src/main/scala/com/spotify/scio/coders/BeamCoders.scala

```diff
@@ -17,110 +17,91 @@
+    val options = CoderOptions(coll.context.options)
     val coder = coll.internal.getCoder
-    val (a, b, c) = unwrap(coder) match {
-      case c: scio.Tuple3Coder[A, B, C] => (c.ac, c.bc, c.cc)
-      case c: RecordCoder[(A, B, C)] =>
-        val ac = coderElement[A](c)(0)
-        val bc = coderElement[B](c)(1)
-        val cc = coderElement[C](c)(2)
-        (ac, bc, cc)
-      case _ =>
+    Some(unwrap(options, coder))
+      .collect { case c: StructuredCoder[_] => c }
+      .map(_.getComponents.asScala.toList)
+      .collect { case (c1: BCoder[A]) :: (c2: BCoder[B]) :: (c3: BCoder[C]) :: Nil =>
+        val a = Coder.beam(unwrap(options, c1))
+        val b = Coder.beam(unwrap(options, c2))
+        val c = Coder.beam(unwrap(options, c3))
+        (a, b, c)
+      }
+      .getOrElse {
         throw new IllegalArgumentException(
           s"Failed to extract tupled coders from Coder[(A, B, C)]: $coder"
         )
-    }
-    (Coder.beam(unwrap(a)), Coder.beam(unwrap(b)), Coder.beam(unwrap(c)))
+      }
   }
 
   def getTuple4Coders[A, B, C, D](
     coll: SCollection[(A, B, C, D)]
   ): (Coder[A], Coder[B], Coder[C], Coder[D]) = {
+    val options = CoderOptions(coll.context.options)
     val coder = coll.internal.getCoder
-    val (a, b, c, d) = unwrap(coder) match {
-      case c: scio.Tuple4Coder[A, B, C, D] => (c.ac, c.bc, c.cc, c.dc)
-      case c: RecordCoder[(A, B, C, D)] =>
-        val ac = coderElement[A](c)(0)
-        val bc = coderElement[B](c)(1)
-        val cc = coderElement[C](c)(2)
-        val dc = coderElement[D](c)(3)
-        (ac, bc, cc, dc)
-      case _ =>
+    Some(unwrap(options, coder))
+      .collect { case c: StructuredCoder[_] => c }
+      .map(_.getComponents.asScala.toList)
+      .collect {
+        case (c1: BCoder[A]) :: (c2: BCoder[B]) :: (c3: BCoder[C]) :: (c4: BCoder[D]) :: Nil =>
+          val a = Coder.beam(unwrap(options, c1))
+          val b = Coder.beam(unwrap(options, c2))
+          val c = Coder.beam(unwrap(options, c3))
+          val d = Coder.beam(unwrap(options, c4))
+          (a, b, c, d)
+      }
+      .getOrElse {
         throw new IllegalArgumentException(
           s"Failed to extract tupled coders from Coder[(A, B, C, D)]: $coder"
         )
-    }
-    (Coder.beam(unwrap(a)), Coder.beam(unwrap(b)), Coder.beam(unwrap(c)), Coder.beam(unwrap(d)))
-  }
-
-  private def getIterableV[V](coder: beam.Coder[Iterable[V]]): beam.Coder[V] =
-    unwrap(coder) match {
-      case c: scio.BaseSeqLikeCoder[Iterable, V] @unchecked => c.elemCoder
-      case _ =>
-        throw new IllegalArgumentException(
-          s"Failed to extract value coder from Coder[Iterable[V]]: $coder"
-        )
-    }
-
-  /** Get key-value coders from a `SideInput[Map[K, Iterable[V]]]`. */
-  def getMultiMapKV[K, V](si: SideInput[Map[K, Iterable[V]]]): (Coder[K], Coder[V]) = {
-    val coder = si.view.getPCollection.getCoder
```

Comment:
This was very unsafe. The `PCollection` behind a view has no guarantee to be of the same type (int contains a transform function). Drop this helper and expect a coder when we pass a `SideInput`